lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// ErrFwdNotExists is an error returned when the caller tries to resolve
+	// a forward that doesn't exist anymore.
+	ErrFwdNotExists = errors.New("forward does not exist")
+
+	// ErrMissingPreimage is an error returned when the caller tries to settle
+	// a forward and doesn't provide a preimage.
+	ErrMissingPreimage = errors.New("missing preimage")
+
+	// ErrInterceptorAlreadyExists is an error returned when the a new stream
+	// is opened and there is already one active interceptor.
+	// The user must disconnect prior to open another stream.
+	ErrInterceptorAlreadyExists = errors.New("interceptor already exists")
+)
+
+// HtlcInterceptor is a bidirectional stream for streaming interception
+// requests to the caller.
+// Upon connection it does the following:
+// 1. Check if there is already a live stream, if yes it rejects the request.
+// 2. Regsitered a ForwardInterceptor
+// 3. Delivers to the caller every √√ and detect his answer.
+// It uses a local implementation of holdForwardsStore to keep all the hold
+// forwards and find them when manual resolution is later needed.
+func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
+	// We ensure there is only one interceptor at a time.
+	rpcInterceptor := newRPCInterceptor(stream)
+	if err := s.setRPCInterceptor(rpcInterceptor); err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	quit := make(chan struct{})
+	defer func() {
+		close(quit)
+		s.onInterceptorDisconnected(rpcInterceptor)
+		wg.Wait()
+	}()
+
+	// this go routine reads from the client stream and and then signal to the
+	// main loop to handle the packet.
+	clientResolutionsChan := make(chan *ForwardHtlcInterceptResponse)
+	errChan := make(chan error)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			// Now that we have the response from the RPC client, send it to
+			// the responses chan.
+			select {
+			case clientResolutionsChan <- resp:
+			case <-quit:
+				return
+			case <-s.quit:
+				return
+			}
+		}
+	}()
+
+	// Our interceptor makes sure we hold the packet and then signal to the
+	// main loop to handle the packet. We only return true if we were able
+	// to deliver the packet to the main loop.
+	interceptedPacketsChan := make(chan htlcswitch.InterceptedForward)
+	fromSwitchInterceptor := func(packet htlcswitch.InterceptedForward) bool {
+		log.Infof("holding intercepted packet %v", packet)
+		rpcInterceptor.hold(packet)
+		select {
+		case interceptedPacketsChan <- packet:
+			return true
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+	}
+
+	// Register our interceptor and make sure we resolve the held packets
+	// on disconnect.
+	interceptableForwarder := s.cfg.RouterBackend.InterceptableForwarder
+	interceptableForwarder.SetInterceptor(fromSwitchInterceptor)
+
+	for {
```

Comment:
Perhaps the main loop would also work well as a method on `rpcInterceptor`

Reply:
I understand the motivation and actually tried that initially and TBH it didn't look better. The function selects from 4 channels that the underline `rpcInterceptor` would need access to...